### PR TITLE
fix(components): [menu] resize event infinite invoke

### LINKED ISSUE #9339: [Component] [menu, menu-item] Menu 菜单，在某些特殊宽度下，会发生最后一个元素不断渲染、删除的问题，造成整个菜单组件不断闪烁
Bug Type: **`Component`**

## Environment

- Vue Version: `3.2.0`
- Element Plus Version: `2.2.11`
- Browser / OS: `Chrome 100.0.4896.88`
- Build Tool: `Webpack`

### Related Component

- `el-menu`
- `el-menu-item`

### Reproduction Link

[Element Plus Playground](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtbWVudVxuICAgIDpkZWZhdWx0LWFjdGl2ZT1cImFjdGl2ZUluZGV4MlwiXG4gICAgY2xhc3M9XCJlbC1tZW51LWRlbW9cIlxuICAgIG1vZGU9XCJob3Jpem9udGFsXCJcbiAgICBiYWNrZ3JvdW5kLWNvbG9yPVwiIzU0NWM2NFwiXG4gICAgdGV4dC1jb2xvcj1cIiNmZmZcIlxuICAgIGFjdGl2ZS10ZXh0LWNvbG9yPVwiI2ZmZDA0YlwiXG4gICAgQHNlbGVjdD1cImhhbmRsZVNlbGVjdFwiXG4gICAgICAgICAgIHN0eWxlPVwid2lkdGg6IDIxNHB4O1wiXG4gID5cbiAgICA8ZWwtbWVudS1pdGVtIGluZGV4PVwiMVwiPlByb2Nlc3NpbmcgQ2VudGVyPC9lbC1tZW51LWl0ZW0+XG4gICAgPGVsLXN1Yi1tZW51IGluZGV4PVwiMlwiPlxuICAgICAgPHRlbXBsYXRlICN0aXRsZT5Xb3Jrc3BhY2U8L3RlbXBsYXRlPlxuICAgICAgPGVsLW1lbnUtaXRlbSBpbmRleD1cIjItMVwiPml0ZW0gb25lPC9lbC1tZW51LWl0ZW0+XG4gICAgICA8ZWwtbWVudS1pdGVtIGluZGV4PVwiMi0yXCI+aXRlbSB0d288L2VsLW1lbnUtaXRlbT5cbiAgICAgIDxlbC1tZW51LWl0ZW0gaW5kZXg9XCIyLTNcIj5pdGVtIHRocmVlPC9lbC1tZW51LWl0ZW0+XG4gICAgICA8ZWwtc3ViLW1lbnUgaW5kZXg9XCIyLTRcIj5cbiAgICAgICAgPHRlbXBsYXRlICN0aXRsZT5pdGVtIGZvdXI8L3RlbXBsYXRlPlxuICAgICAgICA8ZWwtbWVudS1pdGVtIGluZGV4PVwiMi00LTFcIj5pdGVtIG9uZTwvZWwtbWVudS1pdGVtPlxuICAgICAgICA8ZWwtbWVudS1pdGVtIGluZGV4PVwiMi00LTJcIj5pdGVtIHR3bzwvZWwtbWVudS1pdGVtPlxuICAgICAgICA8ZWwtbWVudS1pdGVtIGluZGV4PVwiMi00LTNcIj5pdGVtIHRocmVlPC9lbC1tZW51LWl0ZW0+XG4gICAgICA8L2VsLXN1Yi1tZW51PlxuICAgIDwvZWwtc3ViLW1lbnU+XG4gICAgPGVsLW1lbnUtaXRlbSBpbmRleD1cIjNcIiBkaXNhYmxlZD5JbmZvPC9lbC1tZW51LWl0ZW0+XG4gICAgPGVsLW1lbnUtaXRlbSBpbmRleD1cIjRcIj5PcmRlcnM8L2VsLW1lbnUtaXRlbT5cbiAgPC9lbC1tZW51PlxuPC90ZW1wbGF0ZT5cblxuPHNjcmlwdCBsYW5nPVwidHNcIiBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gJ3Z1ZSdcblxuY29uc3QgYWN0aXZlSW5kZXggPSByZWYoJzEnKVxuY29uc3QgYWN0aXZlSW5kZXgyID0gcmVmKCcxJylcbmNvbnN0IGhhbmRsZVNlbGVjdCA9IChrZXk6IHN0cmluZywga2V5UGF0aDogc3RyaW5nW10pID0+IHtcbiAgY29uc29sZS5sb2coa2V5LCBrZXlQYXRoKVxufVxuPC9zY3JpcHQ+XG4iLCJpbXBvcnRfbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7fVxufSIsIl9vIjp7fX0=)

### Steps to reproduce

菜单的width为215px是能正常显示，但修改为214时则不断闪烁。在实际应用过程中发现，在某些分辨率下会出现菜单闪烁的问题。

### What is Expected?

菜单宽度自适应父级宽度，解决不断闪烁的问题。

### What is actually happening?

某些宽度下，菜单最后一个元素不断渲染、删除，造成菜单闪烁

### Additional comments

_(empty)_

### PR DESCRIPTION
**Pull request by holazz**
closes #9339

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by xiaoxian521** [comment]
`entries` is undefined
image

**Comment by holazz** [comment]
> `entries` is undefined image

Hi, could you give me a hint on how to reproduce this? BTW this PR isn't ready to be reviewed yet so I need to convert it to draft.

**Comment by xiaoxian521** [comment]
打印entries

**Comment by xiaoxian521** [comment]
将下面代码放到`play/src/App.vue`，运行`pnpm dev`即可复现
```
<template>
  <el-menu :default-active="activeIndex" class="el-menu-demo" mode="horizontal" @select="handleSelect">
    <el-menu-item index="1">Processing Center</el-menu-item>
    <el-sub-menu index="2">
      <template #title>Workspace</template>
      <el-menu-item index="2-1">item one</el-menu-item>
      <el-menu-item index="2-2">item two</el-menu-item>
      <el-menu-item index="2-3">item three</el-menu-item>
      <el-sub-menu index="2-4">
        <template #title>item four</template>
        <el-menu-item index="2-4-1">item one</el-menu-item>
        <el-menu-item index="2-4-2">item two</el-menu-item>
        <el-menu-item index="2-4-3">item three</el-menu-item>
      </el-sub-menu>
    </el-sub-menu>
    <el-menu-item index="3" disabled>Info</el-menu-item>
    <el-menu-item index="4">Orders</el-menu-item>
    <el-menu-item index="5">Processing Center</el-menu-item>
    <el-menu-item index="6">Processing Center</el-menu-item>
    <el-menu-item index="7">Processing Center</el-menu-item>
    <el-menu-item index="8">Processing Center</el-menu-item>
    <el-menu-item index="9">Processing Center</el-menu-item>
    <el-menu-item index="10">Processing Center</el-menu-item>
    <el-menu-item index="11">Processing Center</el-menu-item>
    <el-menu-item index="12">Processing Center</el-menu-item>
  </el-menu>
</template>

<script lang="ts" setup>
import { ref } from 'vue'

const activeIndex = ref('1')
const handleSelect = (key: string, keyPath: string[]) => {
  console.log(key, keyPath)
}
</script>

<style>
.el-menu-demo {
  min-width: 0;
  flex: 1;
}
</style>
```

**Comment by holazz** [comment]
> 将下面代码放到`play/src/App.vue`，运行`pnpm dev`即可复现
> 
> ```
> <template>
>   <el-menu :default-active="activeIndex" class="el-menu-demo" mode="horizontal" @select="handleSelect">
>     <el-menu-item index="1">Processing Center</el-menu-item>
>     <el-sub-menu index="2">
>       <template #title>Workspace</template>
>       <el-menu-item index="2-1">item one</el-menu-item>
>       <el-menu-item index="2-2">item two</el-menu-item>
>       <el-menu-item index="2-3">item three</el-menu-item>
>       <el-sub-menu index="2-4">
>         <template #title>item four</template>
>         <el-menu-item index="2-4-1">item one</el-menu-item>
>         <el-menu-item index="2-4-2">item two</el-menu-item>
>         <el-menu-item index="2-4-3">item three</el-menu-item>
>       </el-sub-menu>
>     </el-sub-menu>
>     <el-menu-item index="3" disabled>Info</el-menu-item>
>     <el-menu-item index="4">Orders</el-menu-item>
>     <el-menu-item index="5">Processing Center</el-menu-item>
>     <el-menu-item index="6">Processing Center</el-menu-item>
>     <el-menu-item index="7">Processing Center</el-menu-item>
>     <el-menu-item index="8">Processing Center</el-menu-item>
>     <el-menu-item index="9">Processing Center</el-menu-item>
>     <el-menu-item index="10">Processing Center</el-menu-item>
>     <el-menu-item index="11">Processing Center</el-menu-item>
>     <el-menu-item index="12">Processing Center</el-menu-item>
>   </el-menu>
> </template>
> 
> <script lang="ts" setup>
> import { ref } from 'vue'
> 
> const activeIndex = ref('1')
> const handleSelect = (key: string, keyPath: string[]) => {
>   console.log(key, keyPath)
> }
> </script>
> 
> <style>
> .el-menu-demo {
>   min-width: 0;
>   flex: 1;
> }
> </style>
> ```

奇怪，我试了好几次也没有出现 entries undefined 的情况 🤔
WX20220817-112618@2x

**Comment by xiaoxian521** [comment]
但是可以注意到 ... 不见了😂

**Comment by holazz** [comment]
> 但是可以注意到 ... 不见了😂

... 在这里，应该还是有些 bug
WX20220817-114000@2x

**Comment by xiaoxian521** [comment]
Hey, after testing there are two more questions
1. Lost activation effect
image
2. The ellipsis may not be displayed
image

**Comment by holazz** [comment]
@xiaoxian521 谢谢细心 review，我那个判断条件应该还有点问题，我稍后看一下

**Comment by xiaoxian521** [comment]
> @xiaoxian521 谢谢细心 review，我那个判断条件应该还有点问题，我稍后看一下

好的 如果好了@我一下 我会继续测试😊

**Comment by holazz** [comment]
> > @xiaoxian521 谢谢细心 review，我那个判断条件应该还有点问题，我稍后看一下
> 
> 
> 
> 好的 如果好了@我一下 我会继续测试😊

好的，改好之后我@你

**Comment by holazz** [comment]
Hi, @xiaoxian521 我觉得可以了，麻烦再审查一下！

File: packages/components/menu/__tests__/menu.test.ts [hunk 1]
```diff
@@ -423,30 +423,6 @@ describe('other', () => {
 
     expect(onOpen).toHaveBeenCalled()
   })
-  test('ellipsis', async () => {
```
Comment (xiaoxian521) [inline packages/components/menu/__tests__/menu.test.ts:426]: 嘿，这个单元测试为什么会删除呢

File: packages/components/menu/src/menu.ts [hunk 2]
```diff
@@ -213,8 +216,33 @@ export default defineComponent({
...
+        if (calcWidth <= menuWidth - moreItemWidth) {
+          sliceIndex = index + 1
+        }
+      })
+      return sliceIndex
+    }
+
     // Common computer monitor FPS is 60Hz, which means 60 redraws per second. Calculation formula: 1000ms/60 ≈ 16.67ms
-    const debounce = (fn: () => void, wait = 16.67) => {
+    const debounce = (fn: () => void, wait = 16.67 * 2) => {
```
Comment (xiaoxian521) [inline packages/components/menu/src/menu.ts:245]: 这个`wait`从16.67变成16.67*2的考虑是什么呢

File: packages/components/menu/__tests__/menu.test.ts (hunk 1 above)
Comment (holazz) [reply packages/components/menu/__tests__/menu.test.ts:426]: 因为 jsdom 不支持，所以无法测试，之前测试通过只是碰巧

WX20220819-180805@2x

File: packages/components/menu/src/menu.ts (hunk 2 above)
Comment (holazz) [reply packages/components/menu/src/menu.ts:245]: 因为每次 resize 时可能会执行两次渲染

WX20220819-181139@2x

**Review by xiaoxian521 — APPROVED** [review]

**Review by xiaoxian521 — APPROVED** [review]

**Comment by xiaoxian521** [comment]
@holazz 干得漂亮❤️

**Pull request merged**